Transmit CG display list surfaces out-of-line

hortont424 opened the pull request:
#### 6b84b375f3c1660e48f8396904b1604c89e1fdb6
<pre>
Transmit CG display list surfaces out-of-line
<a href="https://bugs.webkit.org/show_bug.cgi?id=243232">https://bugs.webkit.org/show_bug.cgi?id=243232</a>
&lt;rdar://93852164&gt;

Reviewed by Simon Fraser.

Instead of encoding images into the display list data blob, allow the encoder
to send them as sidecar IOSurfaces.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
Add a preference to allow disabling out-of-line surfaces.

* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::CreationContext::CreationContext):
Add a bit to the ImageBuffer CreationContext to plumb the desire to use out-of-line
surfaces or not through to the backend. This is a slight layering oddity, but
seems to be the current state-of-the-art for passing parameters to backends.

* Source/WebKit/Shared/RemoteLayerTree/CGDisplayList.cpp:
(WebKit::CGDisplayList::encode const):
(WebKit::CGDisplayList::decode):
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayList.h:
(WebKit::CGDisplayList::CGDisplayList):
(WebKit::CGDisplayList::takeSurfaces):
Store and transmit a set of surfaces alongside the display list blob.

* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.mm: Renamed from Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.cpp.
(WTF::CFTypeTrait&lt;CAMachPortRef&gt;::typeID):
(WebKit::GraphicsContextCGDisplayList::GraphicsContextCGDisplayList):
(WebKit::CGDisplayListImageBufferBackend::calculateMemoryCost):
(WebKit::CGDisplayListImageBufferBackend::create):
(WebKit::CGDisplayListImageBufferBackend::CGDisplayListImageBufferBackend):
(WebKit::CGDisplayListImageBufferBackend::createBackendHandle const):
If enabled, ask the encoder to store IOSurfaces out-of-line, and stick them in a
Vector of MachSendRights ready for IPC.

(WebKit::CGDisplayListImageBufferBackend::context const):
(WebKit::CGDisplayListImageBufferBackend::backendSize const):
(WebKit::CGDisplayListImageBufferBackend::bytesPerRow const):
(WebKit::CGDisplayListImageBufferBackend::copyNativeImage const):
(WebKit::CGDisplayListImageBufferBackend::getPixelBuffer const):
(WebKit::CGDisplayListImageBufferBackend::putPixelBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::ensureBackingStore):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::applyBackingStoreToLayer):
Because of the MachSendRights, we now actually steal the m_displayListBufferHandle
when applying backing store to a layer. This is not a problem in practice
because backing store is only applied to a layer once.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.h:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.mm:
(-[WKCompositingLayer _setWKContents:withDisplayList:replayForTesting:]):
Convert the MachSendRights back into CAMachPorts and hand them off to the layer.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.cpp:
(WebKit::PlatformCALayerRemote::updateBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
(WebKit::RemoteLayerTreeContext::useCGDisplayListOutOfLineSurfaces const):
(WebKit::RemoteLayerTreeContext::setUseCGDisplayListOutOfLineSurfaces):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updatePreferences):
Disable out-of-line surfaces if we are replaying display lists ourselves
(because we have no API to replay with a set of surfaces), or if the setting
is disabled.

Canonical link: <a href="https://commits.webkit.org/252874@main">https://commits.webkit.org/252874@main</a>
</pre>

smfr commented on Source/WebKit/Shared/RemoteLayerTree/CGDisplayList.h:
```diff
@@ -35,21 +35,24 @@ class CGDisplayList {
 public:
     CGDisplayList() = default;
-    explicit CGDisplayList(WebCore::SharedBuffer& displayList)
+    CGDisplayList(WebCore::SharedBuffer& displayList, Vector<MachSendRight>&& surfaces)
```
Maybe we should make a type for a MachSendRight that represents a surface.

hortont424 replied:
You mean just a using/typedef?

smfr commented on Source/WebKit/Shared/RemoteLayerTree/CGDisplayList.h:
```diff
@@ -35,21 +35,24 @@ class CGDisplayList {
     CGDisplayList& operator=(CGDisplayList&&) = default;
 
     RefPtr<WebCore::SharedBuffer> buffer() const { return m_displayList; }
+    Vector<MachSendRight>&& takeSurfaces() { return WTFMove(m_surfaces); }
```
WTFMove or `std::exchange` I feel like the latter is better.

hortont424 replied:
True, good to leave something sane behind.

cdumez replied:
This function is not good, it doesn't take anything. It allowed the caller to take.

Should either be:
```
Vector<MachSendRight> takeSurfaces() { return WTFMove(m_surfaces); }
```
or
```
Vector<MachSendRight>&& takeSurfaces() { return std::exchange(m_surfaces, Vector<MachSendRight>()); }
```

I recommend the former.

cdumez commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -59,18 +59,28 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThr
         bool avoidIOSurfaceSizeCheckInWebProcessForTesting = false;
 
+#if ENABLE(CG_DISPLAY_LIST_BACKED_IMAGE_BUFFER)
+        enum class UseOutOfLineSurfaces : uint8_t { No, Yes };
```
Why uint8_t and not bool?

cdumez commented on Source/WebKit/Shared/RemoteLayerTree/CGDisplayList.h:
```diff
@@ -35,21 +35,24 @@ class CGDisplayList {
-    explicit CGDisplayList(WebCore::SharedBuffer& displayList)
+    CGDisplayList(WebCore::SharedBuffer& displayList, Vector<MachSendRight>&& surfaces)
         : m_displayList(&displayList)
+        , m_surfaces(surfaces)
```
Missing WTFMove() ?

cdumez commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -366,8 +367,11 @@
+    if (m_includeDisplayList == IncludeDisplayList::Yes) {
+        ImageBuffer::CreationContext creationContext;
+        creationContext.useOutOfLineSurfacesForCGDisplayLists = m_useOutOfLineSurfaces;
+        m_frontBuffer.displayListImageBuffer = ImageBuffer::create<CGDisplayListImageBufferBackend>(m_size, m_scale, DestinationColorSpace::SRGB(), pixelFormat(), RenderingPurpose::DOM, creationContext);
```
We may want to WTFMove(creationContext).

hortont424 committed:
Transmit CG display list surfaces out-of-line
https://bugs.webkit.org/show_bug.cgi?id=243232
<rdar://93852164>

Reviewed by Simon Fraser.

Instead of encoding images into the display list data blob, allow the encoder
to send them as sidecar IOSurfaces.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
Add a preference to allow disabling out-of-line surfaces.

* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::CreationContext::CreationContext):
Add a bit to the ImageBuffer CreationContext to plumb the desire to use out-of-line
surfaces or not through to the backend. This is a slight layering oddity, but
seems to be the current state-of-the-art for passing parameters to backends.

* Source/WebKit/Shared/RemoteLayerTree/CGDisplayList.cpp:
(WebKit::CGDisplayList::encode const):
(WebKit::CGDisplayList::decode):
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayList.h:
(WebKit::CGDisplayList::CGDisplayList):
(WebKit::CGDisplayList::takeSurfaces):
Store and transmit a set of surfaces alongside the display list blob.

* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.mm: Renamed from Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.cpp.
(WTF::CFTypeTrait<CAMachPortRef>::typeID):
(WebKit::GraphicsContextCGDisplayList::GraphicsContextCGDisplayList):
(WebKit::CGDisplayListImageBufferBackend::calculateMemoryCost):
(WebKit::CGDisplayListImageBufferBackend::create):
(WebKit::CGDisplayListImageBufferBackend::CGDisplayListImageBufferBackend):
(WebKit::CGDisplayListImageBufferBackend::createBackendHandle const):
If enabled, ask the encoder to store IOSurfaces out-of-line, and stick them in a
Vector of MachSendRights ready for IPC.

(WebKit::CGDisplayListImageBufferBackend::context const):
(WebKit::CGDisplayListImageBufferBackend::backendSize const):
(WebKit::CGDisplayListImageBufferBackend::bytesPerRow const):
(WebKit::CGDisplayListImageBufferBackend::copyNativeImage const):
(WebKit::CGDisplayListImageBufferBackend::getPixelBuffer const):
(WebKit::CGDisplayListImageBufferBackend::putPixelBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::ensureBackingStore):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::applyBackingStoreToLayer):
Because of the MachSendRights, we now actually steal the m_displayListBufferHandle
when applying backing store to a layer. This is not a problem in practice
because backing store is only applied to a layer once.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.h:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.mm:
(-[WKCompositingLayer _setWKContents:withDisplayList:replayForTesting:]):
Convert the MachSendRights back into CAMachPorts and hand them off to the layer.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.cpp:
(WebKit::PlatformCALayerRemote::updateBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
(WebKit::RemoteLayerTreeContext::useCGDisplayListOutOfLineSurfaces const):
(WebKit::RemoteLayerTreeContext::setUseCGDisplayListOutOfLineSurfaces):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updatePreferences):
Disable out-of-line surfaces if we are replaying display lists ourselves
(because we have no API to replay with a set of surfaces), or if the setting
is disabled.

Canonical link: https://commits.webkit.org/252874@main
